// styletron-engine-atomic/dist-browser-cjs/cache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultiCache = exports.Cache = void 0;

var _sequentialIdGenerator = _interopRequireDefault(require("./sequential-id-generator"));

var _sortCssMediaQueries = _interopRequireDefault(require("./sort-css-media-queries"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class MultiCache {
  constructor(idGenerator, onNewCache, onNewValue) {
    this.idGenerator = idGenerator;
    this.onNewCache = onNewCache;
    this.onNewValue = onNewValue;
    this.sortedCacheKeys = [];
    this.caches = {};
  }

  getCache(key) {
    if (!this.caches[key]) {
      const cache = new Cache(this.idGenerator, this.onNewValue);
      cache.key = key;
      this.sortedCacheKeys.push(key);
      this.sortedCacheKeys.sort(_sortCssMediaQueries.default);
      const keyIndex = this.sortedCacheKeys.indexOf(key);
      const insertBeforeMedia = keyIndex < this.sortedCacheKeys.length - 1 ? this.sortedCacheKeys[keyIndex + 1] : void 0;
      this.caches[key] = cache;
      this.onNewCache(key, cache, insertBeforeMedia);
    }

    return this.caches[key];
  }

  getSortedCacheKeys() {
    return this.sortedCacheKeys;
  }

}

exports.MultiCache = MultiCache;

class Cache {
  constructor(idGenerator, onNewValue) {
    this.cache = {};
    this.idGenerator = idGenerator;
    this.onNewValue = onNewValue;
  }

  addValue(key, value) {
    const cached = this.cache[key];

    if (cached) {
      return cached;
    }

    const id = this.idGenerator.next();
    this.cache[key] = id;
    this.onNewValue(this, id, value);
    return id;
  }

}

exports.Cache = Cache;
