// styletron-engine-atomic/dist-browser-cjs/css.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.atomicSelector = atomicSelector;
exports.declarationsToBlock = declarationsToBlock;
exports.fontFaceBlockToRule = fontFaceBlockToRule;
exports.keyframesBlockToRule = keyframesBlockToRule;
exports.keyframesToBlock = keyframesToBlock;
exports.styleBlockToRule = styleBlockToRule;

var _hyphenateStyleName = _interopRequireDefault(require("./hyphenate-style-name"));

var _validateKeyframesObject = _interopRequireDefault(require("./validate-keyframes-object"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function atomicSelector(id, pseudo) {
  let selector = `.${id}`;

  if (pseudo) {
    selector += pseudo;
  }

  return selector;
}

function keyframesToBlock(keyframes) {
  if (process.env.NODE_ENV !== "production") {
    (0, _validateKeyframesObject.default)(keyframes);
  }

  if (process.env.NODE_ENV !== "production" && typeof Object.getPrototypeOf(keyframes) !== "undefined") {
    if (Object.getPrototypeOf(keyframes) !== Object.getPrototypeOf({})) {
      // eslint-disable-next-line no-console
      console.warn("Only plain objects should be used as animation values. Unexpectedly recieved:", keyframes);
    }
  }

  let result = "";

  for (const animationState in keyframes) {
    result += `${animationState}{${declarationsToBlock(keyframes[animationState])}}`;
  }

  return result;
}

function declarationsToBlock(style) {
  let css = "";

  for (const prop in style) {
    const val = style[prop];

    if (typeof val === "string" || typeof val === "number") {
      css += `${(0, _hyphenateStyleName.default)(prop)}:${val};`;
    }
  } // trim trailing semicolon


  return css.slice(0, -1);
}

function keyframesBlockToRule(id, block) {
  return `@keyframes ${id}{${block}}`;
}

function fontFaceBlockToRule(id, block) {
  return `@font-face{font-family:${id};${block}}`;
}

function styleBlockToRule(selector, block) {
  return `${selector}{${block}}`;
}
